src/nvim/runtime.c

```diff
@@ -269,9 +269,25 @@ void set_context_in_runtime_cmd(expand_T *xp, const char *arg)
   xp->xp_pattern = (char *)arg;
 }
 
-static void source_callback(char *fname, void *cookie)
+static void source_callback(int num_fnames, char **fnames, void *cookie)
 {
-  (void)do_source(fname, false, DOSO_NONE, cookie);
+  for (int i = 0; i < num_fnames; i++) {
+    if (str_ends_with(fnames[i], ".vim")) {
+      (void)do_source(fnames[i], false, DOSO_NONE, cookie);
+    }
+  }
+
+  for (int i = 0; i < num_fnames; i++) {
+    if (str_ends_with(fnames[i], ".lua")) {
+      (void)do_source(fnames[i], false, DOSO_NONE, cookie);
+    }
+  }
+
+  for (int i = 0; i < num_fnames; i++) {
+    if (!str_ends_with(fnames[i], ".vim") && !str_ends_with(fnames[i], ".lua")) {
+      (void)do_source(fnames[i], false, DOSO_NONE, cookie);
+    }
+  }
```

Comment:
This may change behavior if someone uses `:runtime!` to load alphabetically-sorted files (e.g. `00_foo.lua`, `10_foo.vim`, `20_foo.lua`, `30_foo.vim`). Not sure if anyone is using this with multiple extensions though.

Reply:
Ah yes. This changes the behavior of `:runtime! plugin/**/*.\(vim\|lua\)` to be the same of `:runtime! plugin/**/*.vim plugin/**/*.lua`.

Reply:
I assumed that was intentional. Otherwise all of these `do_source` calls could be in a single for-loop.

Reply:
I think this behavior change needs to be documented. To achieve the old behavior one has to expand the pattern and the source the files one by one.

Reply:
In any case, we should decide and unambiguously document the order. (Which this PR is already improving.)

Reply:
We don't want to lock down the implementation too much without good reason. I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`. Anything outside of that should be explicitly undefined.

Reply:
> I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`.

This should still be documented at `:h runtime`, as previously `:runtime` doesn't have such special-cased behavior.

Reply:
> We don't want to lock down the implementation too much without good reason. I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`. Anything outside of that should be explicitly undefined.

Then we should document _that_. But I do think it's important to know whether `foo/bar.lua` is loaded before `foo/baz.vim` or not.

Reply:
Added something to `:h runtime`.